File: Python/split-two-strings-to-make-palindrome.py
# Time:  O(n)
# Space: O(1)

class Solution(object):
    def checkPalindromeFormation(self, a, b):
        """
        :type a: str
        :type b: str
        :rtype: bool
        """
        def is_palindrome(s, i, j):
            while i < j:
                if s[i] != s[j]:
                    return False
                i += 1
                j -= 1
            return True

        def check(a, b):
            i, j = 0, len(b)-1
            while i < j:
                if a[i] != b[j]:
                    return is_palindrome(a, i, j) or is_palindrome(b, i, j)
                i += 1
                j -= 1
            return True

        return check(a, b) or check(b, a)

File: C++/split-two-strings-to-make-palindrome.cpp
// Time:  O(n)
// Space: O(1)

class Solution {
public:
    bool checkPalindromeFormation(string a, string b) {
        return check(a, b) || check(b, a);
    }

private:
    bool isPalindrome(const string &s, int i, int j) {
        for (; i < j; ++i, --j) {
            if (s[i] != s[j]) {
                return false;
            }
        }
        return true;
    }
    
    bool check(const string &a, const string &b) {
        for (int i = 0, j = size(b) - 1; i < j; ++i, --j) {
            if (a[i] != b[j]) {
                return isPalindrome(a, i, j) || isPalindrome(b, i, j);
            }
        }
        return true;
    }
    
};
